### PR TITLE
Sdk dict module update

### PR DESCRIPTION
**Pull request by osborneMs**

### CODE REVIEW

File: morphir/sdk/core/src/morphir/sdk/Dict.scala [hunk 1]
```diff
@@ -19,16 +19,36 @@ package morphir.sdk
...
+    }
+
+  def update[K, V](targetKey: K)(updatedValue: V)(dict: Dict[K, V]): Dict[K, V] = dict match {
+    case EmptyDict         => dict
+    case DictImpl(wrapped) => DictImpl(wrapped.updated(targetKey, updatedValue))
+  }
+
+  def remove[K, V](targetKey: K)(dict: Dict[K, V]): Dict[K, V] = dict match {
+    case EmptyDict         => dict
+    case DictImpl(wrapped) => DictImpl(wrapped.filterKeys(key => key != targetKey))
```
Comment (AttilaMihaly) [inline morphir/sdk/core/src/morphir/sdk/Dict.scala:44]: The dictionary might become empty after the filtering so we need an additional check here and return EmptyDict in that case.

File: morphir/sdk/core/src/morphir/sdk/Dict.scala [hunk 2]
```diff
@@ -60,14 +93,58 @@ object Dict {
...
     }
 
-  def toList[K, V](dict: Dict[K, V]): List[(K, V)] =
+  /* Transform */
+  def map[K, V, B](f: K => V => B)(dict: Dict[K, V]): Dict[K, B] = dict match {
+    case EmptyDict         => empty
+    case DictImpl(wrapped) => DictImpl(wrapped.map(x => (x._1, f(x._1)(x._2))))
+  }
+
+  def foldl[K, V, B](initValue: B)(f: B => (K, V) => B)(dict: Dict[K, V]): B = dict match {
```
Comment (AttilaMihaly) [inline morphir/sdk/core/src/morphir/sdk/Dict.scala:102]: This type signature doesn't look right. The Elm version looks like this:
```elm
foldl : (k -> v -> b -> b) -> b -> Dict k v -> b
```
So the direct translation is:
```scala
def foldl[K, V, B](f: K => V => B => B)(initValue: B)(dict: Dict[K, V]): B
```

File: morphir/sdk/core/src/morphir/sdk/Dict.scala [hunk 3]
```diff
@@ -60,14 +93,58 @@ object Dict {
...
+    case EmptyDict => initValue
+    case DictImpl(wrapped) =>
+      wrapped.foldLeft(initValue)((accumulator, pairedValues) =>
+        f(accumulator)(pairedValues._1, pairedValues._2)
+      )
+  }
+
+  def foldr[K, V, B](initValue: B)(f: (K, V) => B => B)(dict: Dict[K, V]): B = dict match {
```
Comment (AttilaMihaly) [inline morphir/sdk/core/src/morphir/sdk/Dict.scala:110]: Same here. Make sure that the type signature matches the Elm version.

File: morphir/sdk/core/src/morphir/sdk/Dict.scala [hunk 4]
```diff
@@ -60,14 +93,58 @@ object Dict {
...
+  }
+
+  def filter[K, V](f: K => V => Boolean)(dict: Dict[K, V]): Dict[K, V] = dict match {
+    case EmptyDict => empty
+    case DictImpl(wrapped) =>
+      val filtered = wrapped.filter(x => f(x._1)(x._2))
+      if (filtered.isEmpty) empty else DictImpl(filtered)
+  }
+
+  def partition[K, V](f: ((K, V)) => Boolean)(dict: Dict[K, V]): (Dict[K, V], Dict[K, V]) =
```
Comment (AttilaMihaly) [inline morphir/sdk/core/src/morphir/sdk/Dict.scala:123]: Type signature doesn't match Elm version. `f` needs to be curried.

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: morphir/sdk/core/src/morphir/sdk/Dict.scala (hunk 3 above)
Comment (osborneMs) [reply morphir/sdk/core/src/morphir/sdk/Dict.scala:110]: fixed

File: morphir/sdk/core/src/morphir/sdk/Dict.scala [hunk 5]
```diff
@@ -19,16 +19,41 @@ package morphir.sdk
...
+    case DictImpl(wrapped) => DictImpl(wrapped.updated(targetKey, updatedValue))
+  }
+
+  def remove[K, V](targetKey: K)(dict: Dict[K, V]): Dict[K, V] = dict match {
+    case EmptyDict => dict
+    case DictImpl(wrapped) =>
+      val res = wrapped.filterKeys(key => key != targetKey)
+      res match {
+        case Map.empty => EmptyDict
+        case resultMap => DictImpl(resultMap)
```
Comment (AttilaMihaly) [inline morphir/sdk/core/src/morphir/sdk/Dict.scala:48]: I don't think `Map.empty` will match correctly. Can you turn this into an if/else with `isEmpty` instead?

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]

File: morphir/sdk/core/src/morphir/sdk/Dict.scala (hunk 5 above)
Comment (osborneMs) [reply morphir/sdk/core/src/morphir/sdk/Dict.scala:48]: Noted

**Review by AttilaMihaly — APPROVED** [review]

**Pull request merged**